JMS 2.0 bug (Container managed JMSContext)

Hello everyone,

Looks like something wrong with connections in JMS 2.0

Using JMS 2.0 API  we are faced with following exception:

> Caused by: javax.ejb.TransactionRolledbackLocalException: Exception thrown from bean
	at com.sun.ejb.containers.EJBContainerTransactionManager.checkExceptionClientTx(EJBContainerTransactionManager.java:664)
	at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:509)
	at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:4542)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2021)
	... 99 more
Caused by: com.sun.messaging.jms.MQRuntimeException: MQRA:CFA:allocation failure:createConnection:Error in allocating a connection. Cause: java.lang.IllegalStateException: Wrong XAState: 0
	at com.sun.messaging.jmq.jmsclient.JMSContextImpl.<init>(JMSContextImpl.java:230)
	at com.sun.messaging.jms.ra.ConnectionFactoryAdapter.createContext(ConnectionFactoryAdapter.java:182)
	at org.glassfish.jms.injection.AbstractJMSContextManager.createContext(AbstractJMSContextManager.java:80)

You can reproduce with this project: https://github.com/vkalayda/ee-jms-20-bug

**How to reproduce**

* Create JMS Destination Resource - jms/TestMessageQueue
* Build and Deploy project "ee-jms-20-bug"
* Execute following command "ab -n 100 -c 10 http://localhost:8080/ee-jms-20-bug/resources/test/jms20"
 
But with JMS 1.1 we haven't errors

* Execute following command "ab -n 100 -c 10 http://localhost:8080/ee-jms-20-bug/resources/test/jms11"

Another one interesting thing - Removing `@Stateless` annotation from "MessageSender20" (put it under the CDI container) will eliminate exceptions.


Also if each time create and close the JMSContext (Application Managed JMSContext) all works fine.

```
@Resource(mappedName = "jms/TestMessageQueue")
    private Queue testMessageQueue;

    @Resource(lookup = "java:comp/DefaultJMSConnectionFactory")
    private ConnectionFactory connectionFactory;

    public void send() {
        try (JMSContext context = connectionFactory.createContext()) {
            context.createProducer().send(testMessageQueue, "Hello world");
        }
    }
```

From JMSContext documentation:

https://docs.oracle.com/javaee/7/api/javax/jms/JMSContext.html

> A JMSContext may be created by calling one of several createContext methods on a ConnectionFactory. A JMSContext that is created in this way is described as being application-managed. An application-managed JMSContext must be closed when no longer needed by calling its close method.

This works.

> Applications running in the Java EE web and EJB containers may alternatively inject a JMSContext into their application using the `@Inject` annotation. A JMSContext that is created in this way is described as being container-managed. A container-managed JMSContext will be closed automatically by the container.

Doesn't work.


Was reproduced on Full **Payara 4.1.1.164**

--- a/appserver/ejb/ejb-container/src/main/java/com/sun/ejb/containers/EJBContextImpl.java
+++ b/appserver/ejb/ejb-container/src/main/java/com/sun/ejb/containers/EJBContextImpl.java
@@ -37,6 +37,7 @@
  * only if the new code is made subject to such option by the copyright
  * holder.
  */
+// Portions Copyright [2016-2018] [Payara Foundation and/or its affiliates.]
 
 package com.sun.ejb.containers;
 
@@ -44,9 +45,6 @@ import com.sun.appserv.connectors.internal.api.ResourceHandle;
 import com.sun.ejb.ComponentContext;
 import com.sun.ejb.Container;
 import com.sun.ejb.EjbInvocation;
-import com.sun.enterprise.deployment.EjbDescriptor;
-import com.sun.enterprise.deployment.RoleReference;
-import com.sun.enterprise.transaction.api.JavaEETransaction;
 import com.sun.enterprise.util.LocalStringManagerImpl;
 import org.glassfish.api.invocation.ComponentInvocation;
 
@@ -275,7 +273,14 @@ public abstract class EJBContextImpl
             resources = new ArrayList(0);
         return resources;
     }
-    
+
+    /**
+     * reset resource list when putting SLSB back in the pool
+     */
+    void resetResourceList() {
+        resources = null;
+    }
+
     
     /**
      * Get the number of concurrent invocations on this bean

--- a/appserver/ejb/ejb-container/src/main/java/com/sun/ejb/containers/StatelessSessionContainer.java
+++ b/appserver/ejb/ejb-container/src/main/java/com/sun/ejb/containers/StatelessSessionContainer.java
@@ -37,7 +37,7 @@
  * only if the new code is made subject to such option by the copyright
  * holder.
  */
-// Portions Copyright [2016-2017] [Payara Foundation and/or its affiliates]
+// Portions Copyright [2016-2018] [Payara Foundation and/or its affiliates]
 
 package com.sun.ejb.containers;
 
@@ -71,10 +71,8 @@ import com.sun.ejb.monitoring.stats.EjbMonitoringStatsProvider;
 import com.sun.ejb.monitoring.stats.EjbPoolStatsProvider;
 import com.sun.ejb.monitoring.stats.StatelessSessionBeanStatsProvider;
 import com.sun.enterprise.admin.monitor.callflow.ComponentType;
-import com.sun.enterprise.deployment.DescriptorConstants;
 import com.sun.enterprise.deployment.LifecycleCallbackDescriptor.CallbackType;
 import com.sun.enterprise.deployment.runtime.BeanPoolDescriptor;
-import com.sun.enterprise.deployment.xml.RuntimeTagNames;
 import com.sun.enterprise.security.SecurityManager;
 
 /** This class provides container functionality specific to stateless 
@@ -584,6 +582,10 @@ public class StatelessSessionContainer
 
             sc.setState(EJBContextImpl.BeanState.POOLED);
 
+            // prevent java.util.ConcurrentModificationException from JMS
+            // or others that hold on to invocation references, and thus possibly
+            // reference to the contexts that are out of scope
+            sc.resetResourceList();
             // Stateless beans cant have transactions across invocations
             sc.setTransaction(null);
             sc.touch();

